src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,403 @@
+    pub fn warp_sync_peer(&self) -> PeerId {
+        self.state.warp_sync_peer.clone()
+    }
+
+    /// Injects the key.
+    ///
+    /// # Panic
+    ///
+    /// Panics if the key passed as parameter isn't strictly superior to the requested key.
+    ///
+    pub fn inject_key(self, key: Option<impl AsRef<[u8]>>) -> GrandpaWarpSync {
+        GrandpaWarpSync::from_babe_fetch_epoch_query(
+            self.inner.inject_key(key),
+            self.fetched_current_epoch,
+            self.state,
+        )
+    }
+}
+
+/// Verifying the warp sync response is required to continue.
+pub struct Verifier {
+    verifier: warp_sync::Verifier,
+    genesis_chain_information: ChainInformation,
+    warp_sync_peer: PeerId,
+}
+
+impl Verifier {
+    pub fn next(self) -> GrandpaWarpSync {
+        match self.verifier.next() {
+            Ok(warp_sync::Next::NotFinished(next_verifier)) => GrandpaWarpSync::Verifier(Self {
+                verifier: next_verifier,
+                genesis_chain_information: self.genesis_chain_information,
+                warp_sync_peer: self.warp_sync_peer,
+            }),
+            Ok(warp_sync::Next::Success {
+                header,
+                chain_information_finality,
+            }) => GrandpaWarpSync::VirtualMachineParamsGet(VirtualMachineParamsGet {
+                state: PostVerificationState {
+                    header,
+                    chain_information_finality,
+                    genesis_chain_information: self.genesis_chain_information,
+                    warp_sync_peer: self.warp_sync_peer,
+                },
+            }),
+            Err(error) => GrandpaWarpSync::Finished(Err(Error::Verifier(error))),
+        }
+    }
+}
+
+struct PostVerificationState {
+    header: Header,
+    chain_information_finality: ChainInformationFinality,
+    genesis_chain_information: ChainInformation,
+    warp_sync_peer: PeerId,
+}
+
+/// Performing a GrandPa warp sync network request is required to continue.
+pub struct WarpSyncRequest {
+    peer_index: usize,
+    peers: Vec<PeerId>,
+    genesis_chain_information: ChainInformation,
+}
+
+impl WarpSyncRequest {
+    /// The peer to make a GrandPa warp sync network request to.
+    pub fn current_peer(&self) -> PeerId {
+        self.peers[self.peer_index].clone()
+    }
+
+    /// Add a peer to the list of peers.
+    pub fn add_peer(&mut self, peer: PeerId) {
+        assert!(!self.peers.iter().any(|p| p == &peer));
+        self.peers.push(peer);
+    }
+
+    /// Remove a peer from the list of peers.
+    ///
+    /// # Panic
+    ///
+    /// Panics if the peer wasn't added to the list earlier.
+    ///
+    pub fn remove(mut self, to_remove: PeerId) -> GrandpaWarpSync {
+        if to_remove == self.current_peer() {
+            let next_index = self.peer_index + 1;
+
+            if next_index == self.peers.len() {
+                GrandpaWarpSync::WaitingForPeers(WaitingForPeers {
+                    peers: Vec::new(),
```

Comment:
```suggestion
                    peers: self.peers,
```

Reply:
If we do this, then the next call to `WaitingForPeers::add_peer` will reset the `peer_index` to 0. Is it indented that we go through all the bad peers again?